### DISCUSSION TITLE
How does it work

### DISCUSSION THREAD

**Discussion by kkaryos** [autodiff/autodiff#334] [General]
Is published reference for the method that is being used? Basically, how this autodiff library works "under the hood".

**Comment by allanleal** [comment]
Hi @kkaryos - thanks for your question. There is still no publication describing the methodology. Source code is currently the only form of algorithm description. The algorithm behind the “dual” implementation is the standard forward mode with operator overloading relying heavily on template metaprogramming to optimize calculations and avoid temporaries. The implementation of "real" is based on forward mode with derivative expressions designed for higher order directional derivatives. Several people have contributed to these codes in the past with improvements, fixes, and support for new mathematical functions, meaning that the methodology can be inferred from the code, but I agree that this is not the best format of presentation and it would great to have something published in the future.
